File: Windows/src/cloudprism/core/decryptor.py
"""下载解密管线。

Decryptor 负责从存储后端下载 .cpenc 容器并解密为本地明文文件：

    1. 从后端下载文件头（前 64 字节足够覆盖 51B 头）
    2. 解析头获取 salt / iv / header_length
    3. 用 Session 派生密钥
    4. 分块下载密文主体 -> AES-CTR 解密 -> 写本地明文

流式解密无填充，明文长度 = 密文长度 = 文件大小 - header_length。
"""

from __future__ import annotations

import os
from io import BytesIO
from typing import Iterator

from Crypto.Cipher import AES

from cloudprism import constants
from cloudprism.core.session import Session
from cloudprism.crypto.header import FileHeader
from cloudprism.storage.backend import StorageBackend


class Decryptor:
    """解密器：下载 .cpenc 容器 -> 解密 -> 写本地明文。"""

    # 默认分块大小（字节）：1 MiB
    DEFAULT_CHUNK: int = 1 << 20

    def __init__(
        self,
        session: Session,
        backend: StorageBackend,
        chunk: int = DEFAULT_CHUNK,
    ) -> None:
        self.session = session
        self.backend = backend
        self.chunk = chunk

    def _fetch_header(self, remote_path: str) -> FileHeader:
        """取文件头并解析。

        前 64 字节足够覆盖默认 51 字节头（salt16/iv16）。
        """
        head_bytes = self.backend.download_range(remote_path, 0, 63)
        return FileHeader.parse(BytesIO(head_bytes))

    def download_and_decrypt(
        self,
        remote_path: str,
        local_path: str,
    ) -> Iterator[float]:
        """从后端下载并解密到本地，yield 进度 0.0~1.0。

        参数:
            remote_path: 后端上的 .cpenc 文件路径
            local_path: 本地明文输出路径
        """
        # 1. 取文件头
        header = self._fetch_header(remote_path)
        total_cipher = self.backend.get_size(remote_path)
        total_plain = total_cipher - header.header_length

        # 2. 派生密钥
        key = self.session.derive_key(header.salt)

        # 3. 准备 CTR 解密器（与加密用同一计数器方案）
        from Crypto.Util import Counter

        initial_value = int.from_bytes(header.iv, "big")
        ctr = Counter.new(
            128, initial_value=initial_value, allow_wraparound=True
        )
        cipher = AES.new(key, AES.MODE_CTR, counter=ctr)

        # 4. 分块下载密文主体（跳过文件头）-> 解密 -> 写明文
        offset = header.header_length
        written = 0
        with open(local_path, "wb") as f:
            while written < total_plain:
                # 本块要下载的字节数（密文=明文，等长）
                want = min(self.chunk, total_plain - written)
                # 密文文件偏移 = header_length + 已解密字节数
                ct_start = offset + written
                ct_end = ct_start + want - 1
                ct_block = self.backend.download_range(
                    remote_path, ct_start, ct_end
                )
                f.write(cipher.decrypt(ct_block))
                written += len(ct_block)
                yield written / total_plain if total_plain else 1.0

        if total_plain == 0:
            yield 1.0

    def decrypt_range_to_bytes(
        self,
        remote_path: str,
        header: FileHeader | None = None,
        start: int = 0,
        end: int | None = None,
    ) -> bytes:
        """解密远程文件的指定明文区间 [start, end) 为字节。

        供流式代理与测试使用：按需拉取密文块、内存解密、不落盘。
        """
        if header is None:
            header = self._fetch_header(remote_path)

        total_cipher = self.backend.get_size(remote_path)
        total_plain = total_cipher - header.header_length
        if end is None:
            end = total_plain
        if start >= end:
            return b""

        # 用 AesCtrStreamCipher 按块对齐解密（支持随机访问）
        from cloudprism.crypto.stream_cipher import AesCtrStreamCipher

        key = self.session.derive_key(header.salt)
        sc = AesCtrStreamCipher(key, header.iv)
        BS = sc.BLOCK
        first_block = start // BS
        last_block = (end - 1) // BS
        # 密文文件偏移（含 header_length）
        ct_start = header.header_length + first_block * BS
        ct_end = header.header_length + (last_block + 1) * BS - 1
        # end 限制在文件末尾
        ct_end = min(ct_end, total_cipher - 1)
        ct = self.backend.download_range(remote_path, ct_start, ct_end)
        return sc.decrypt_range(ct, first_block, start, end)

File: WindowsGo/pkg/pipeline/decryptor.go
package pipeline

import (
	"context"
	"errors"
	"io"
	"os"
	"sync"
	"sync/atomic"

	"github.com/Sagiri-lzumi/cloudprism/windowsgo/pkg/cryptox"
	"github.com/Sagiri-lzumi/cloudprism/windowsgo/pkg/protocol"
	"github.com/Sagiri-lzumi/cloudprism/windowsgo/pkg/session"
	"github.com/Sagiri-lzumi/cloudprism/windowsgo/pkg/storage"
)

// Decryptor 从存储后端下载 .cpenc 容器并解密。
//
// 两个入口对应 Python Decryptor 的两个方法（decryptor.py）：整文件
// DownloadAndDecrypt（落盘，供传输队列）与 DecryptRangeToBytes（按需
// 拉取明文区间，供缩略图与流式代理），以及 Python 端缩略图/流式依赖的
// 头部随机访问语义 —— 本包只产出明文字节，不感知调用场景。
//
// 对照 WindowsPy/src/cloudprism/core/decryptor.py
type Decryptor struct {
	Sess    *session.Session
	Backend storage.Backend
	Workers int // 整文件下载的解密并发分片数；<=1 或小文件自动退化为单分片
}

// headFetchLen 是取文件头时下载的前缀长度（字节）。
//
// 默认文件头 51 字节（magic8+version4+hl4+saltlen1+salt16+ivlen1+iv16+
// flags1），64 字节足够覆盖 salt/iv 均 16 字节的头（对照 decryptor.py:48）。
const headFetchLen = 64

// fetchHeader 下载文件头并解析。
func (d *Decryptor) fetchHeader(ctx context.Context, remote string) (*cryptox.Header, error) {
	head, err := d.Backend.DownloadRange(ctx, remote, 0, headFetchLen-1)
	if err != nil {
		return nil, err
	}
	return cryptox.ParseBytes(head)
}

// DecryptRangeToBytes 解密远程文件的明文区间 [start, end) 为字节（不落盘）。
//
// end <= 0 表示取到文件末尾。start >= end（或文件无内容）返回空字节串。
// end 超过明文总长时 clamp 到文件尾 —— Python 端靠切片自然截断 + 上层
// 吞 IndexError（decryptor.py:99-134），Go 端切片越界直接 panic，故显式
// clamp，语义等价且不依赖异常。
//
// 供缩略图与流式代理使用：按需拉取密文块、内存解密、不落盘。
func (d *Decryptor) DecryptRangeToBytes(ctx context.Context, remote string, start, end int64) ([]byte, error) {
	if d.Sess == nil || d.Backend == nil {
		return nil, ErrNoSession
	}
	header, err := d.fetchHeader(ctx, remote)
	if err != nil {
		return nil, err
	}
	totalCipher, err := d.Backend.GetSize(ctx, remote)
	if err != nil {
		return nil, err
	}
	totalPlain := PlaintextTotal(int64(header.HeaderLength), totalCipher)
	if end <= 0 || end > totalPlain {
		end = totalPlain
	}
	if start >= end || totalPlain <= 0 {
		return []byte{}, nil
	}

	key, err := d.Sess.DeriveKey(header.Salt)
	if err != nil {
		return nil, err
	}
	ctr, err := cryptox.NewCTR(key[:], header.IV)
	if err != nil {
		return nil, err
	}

	r := PlaintextToCipher(int64(header.HeaderLength), start, end)
	// 密文区间终点 clamp 到文件尾（文件尾块可能不满 16 字节）
	if r.CtEnd > totalCipher {
		r.CtEnd = totalCipher
	}
	ct, err := d.Backend.DownloadRange(ctx, remote, r.CtStart, r.CtEnd-1)
	if err != nil {
		return nil, err
	}
	if len(ct) == 0 {
		return []byte{}, nil
	}

	out := make([]byte, len(ct))
	ctr.XORAt(out, ct, uint64(r.FirstBlock))
	from := start % int64(protocol.BlockSize)
	// 视图切片：明文字节在解密块中的起点（起点处 start 可不对齐 16）
	return out[from : from+(end-start)], nil
}

// DownloadAndDecrypt 整文件下载并解密到本地明文文件。
//
// 进度回调 0.0~1.0（可 nil）。空文件产出空明文并回调一次 1.0。
// 数据量达到 minParallelSize 且 Workers>1 时按 ShardAlign 分片并行：
// 每片一次 Range 请求 + 池缓冲原地解密 + WriterAt 精确落位（os.File 的
// WriteAt 并发安全），顺序由偏移保证，无需收集器；否则退化为单分片顺序。
//
// 注意：失败/取消时**不删除**半成品明文 —— 对齐 Python（decryptor.py:82），
// 由上层（传输队列）决定续传或清理。
func (d *Decryptor) DownloadAndDecrypt(ctx context.Context, remote, local string, report func(float64)) error {
	if d.Sess == nil || d.Backend == nil {
		return ErrNoSession
	}
	if report == nil {
		report = func(float64) {}
	}
	header, err := d.fetchHeader(ctx, remote)
	if err != nil {
		return err
	}
	totalCipher, err := d.Backend.GetSize(ctx, remote)
	if err != nil {
		return err
	}
	totalPlain := PlaintextTotal(int64(header.HeaderLength), totalCipher)
	if totalPlain < 0 {
		return errors.New("pipeline: 远端文件小于文件头，容器损坏")
	}

	f, err := os.Create(local)
	if err != nil {
		return err
	}
	defer f.Close()

	if totalPlain == 0 {
		report(1.0)
		return nil
	}

	key, err := d.Sess.DeriveKey(header.Salt)
	if err != nil {
		return err
	}
	ctr, err := cryptox.NewCTR(key[:], header.IV)
	if err != nil {
		return err
	}
	workers := d.Workers
	if workers < 1 || totalPlain < minParallelSize {
		workers = 1
	}
	return decryptShards(ctx, d.Backend, remote, f, ctr, int64(header.HeaderLength), totalPlain, workers, report)
}

// decryptShards 分片下载解密主循环：workers 个 goroutine 消费分片任务。
//
// workers==1 即顺序路径（同一代码，零调度差异）；每个 goroutine 独占一块
// 1MiB 池缓冲（Range 请求优先走零拷贝直写路径，见 storage.ReadRangeInto），
// 完成后归还复用。进度按已完成字节累计（atomic）。
func decryptShards(ctx context.Context, b storage.Backend, remote string, dst io.WriterAt, ctr *cryptox.CTR, hl, totalPlain int64, workers int, report func(float64)) error {
	shards := Shards(totalPlain)
	if len(shards) == 0 {
		return nil
	}
	if workers > len(shards) {
		workers = len(shards)
	}
	var (
		wg   sync.WaitGroup
		done atomic.Int64
		errs = make(chan error, len(shards))
	)
	jobs := make(chan Shard)
	for w := 0; w < workers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for s := range jobs {
				if err := decryptOneShard(ctx, b, remote, dst, ctr, hl, s, totalPlain, &done, report); err != nil {
					errs <- err
					return // 首错：停止消费，让调度方退出
				}
			}
		}()
	}
feed:
	for _, s := range shards {
		if err := ctx.Err(); err != nil {
			break
		}
		select {
		case jobs <- s:
		case <-ctx.Done():
			break feed
		}
	}
	close(jobs)
	wg.Wait()
	close(errs)
	for err := range errs {
		return err
	}
	return ctx.Err()
}

// decryptOneShard 下载并解密单个分片：Range 请求（含两端）→ 原地 XOR →
// WriterAt 落位到明文偏移。返回错误时进度保持已累计值。
func decryptOneShard(ctx context.Context, b storage.Backend, remote string, dst io.WriterAt, ctr *cryptox.CTR, hl int64, s Shard, totalPlain int64, done *atomic.Int64, report func(float64)) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	buf := cryptox.GetLarge()
	defer cryptox.PutLarge(buf)
	ctStart := hl + s.Offset
	ctEnd := ctStart + s.Length - 1 // Range 含两端
	n, err := storage.ReadRangeInto(ctx, b, remote, ctStart, ctEnd, *buf)
	if err != nil {
		return err
	}
	if int64(n) != s.Length {
		// 数据不足：远端是半成品或后端截断，继续只会错位写坏明文
		return errors.New("pipeline: 远端密文数据不足（容器不完整或已损坏）")
	}
	part := (*buf)[:n]
	ctr.XORAt(part, part, uint64(s.Offset)/uint64(protocol.BlockSize))
	if _, err := dst.WriteAt(part, s.Offset); err != nil {
		return err
	}
	if nd := done.Add(int64(n)); nd > 0 {
		report(float64(nd) / float64(totalPlain))
	}
	return nil
}
